### ISSUE TITLE
Significant Differences in Model Performance Between PMML and actual Model. 

### ISSUE THREAD

**Issue by 1samwatkins3005** [nyoka-pmml/nyoka#29]
I **am noticing significant differences between my normal model, and the pmml of this model. 

Normal Model, where 'VehicleMake' is onehotencoded using OneHotEncoder:** 

enc = OneHotEncoder(handle_unknown='ignore')

enc_df = pd.DataFrame(enc.fit_transform(all1[['VehicleMake']]).toarray())

all1 = all1.join(enc_df)
all1 = all1.drop(columns=['VehicleMake'])
train=all1.sample(frac=0.8,random_state=200)
test=all1.drop(train.index)

#train_repairid = train['RepairID']
train_TL = train['TotalLossIndicator']
train_pred = train.drop(['TotalLossIndicator'], axis=1)

#test_repairid = test['RepairID']
test_TL = test['TotalLossIndicator']
test_pred = test.drop(['TotalLossIndicator'], axis=1)

model = XGBClassifier(max_depth=5, min_child_weight=1, colsample_bytree=1, objective='binary:logistic', n_estimators=50, eta=0.5, n_jobs=7, n_thread=7, scale_pos_weight = 3.1681015821179845, seed=123)

model.fit(train_pred, train_TL)

**PMML Model, doing exactly the same things and using the same parameters as the model above.:** 

target_name='TotalLossIndicator'
pipeline_obj = Pipeline([
    ('mapper', DataFrameMapper([
        (normal_features, None),
        (["VehicleMake"], OneHotEncoder())
    ])),
    ('Xgbc',XGBClassifier(max_depth=5,
                min_child_weight=1,
                colsample_bytree=1,
                objective='binary:logistic',
                n_estimators=50,
                eta=0.5,
                n_jobs=7,
                n_thread=7,
                scale_pos_weight = 3.1681015821179845,
                seed=123
                ))
])

The first model has an accruacy of 88%, the second model is showing an accuracy of 78%. 

Any ideas? Are there any differences between th two models i'm am missing? Or, is there an issue with passing the model to pmml?

**Comment by 1samwatkins3005** [comment]
I've done some further testing using one of your examples for xgboost (with car name omitted for simplicity).  

I ran the following code as a baseline: 

import pandas as pd
from sklearn import datasets
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.feature_extraction.text import CountVectorizer
from xgboost import XGBClassifier,XGBRegressor
from sklearn_pandas import DataFrameMapper
from sklearn.model_selection import train_test_split

auto = pd.read_csv('auto-mpg.csv')
X = auto.drop(['mpg', 'car name'], axis=1)
y = auto['mpg']

x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=101)

model = XGBRegressor(random_state=1, seed=1)

model.fit(x_train, y_train)

---Running predictions with the model above we can see the accuracy with the following plot:

![image](https://user-images.githubusercontent.com/47885698/77522027-7a4b2800-6e7b-11ea-9272-d17679bb1a17.png)


When i run the same model in your wrapper: 

auto = pd.read_csv('auto-mpg.csv')
X = auto.drop(['mpg', 'car name'], axis=1)
y = auto['mpg']

feature_names = [name for name in auto.columns if name not in ('mpg')]

target_name='mpg'


x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=101)
pipeline_obj = Pipeline([
    ('mapper', DataFrameMapper([
        (['displacement'],[StandardScaler()]) 
    ])),
    ('Xgbr',XGBRegressor(random_state=1, seed=1))
])
pipeline_obj.fit(x_train,y_train)

-----I see much less accuracy when predicting:

![image](https://user-images.githubusercontent.com/47885698/77522161-af577a80-6e7b-11ea-9de4-809c0c1e1c7f.png)





Is there any explanation for this? It's quite unexpected?

**Comment by 1samwatkins3005** [comment]
Following some further testing using one of your xgboost examples, i see similar drops in accuracy when wrapping a model in your pmml wrapper. 

Normal model and predictive accuracy: 

import pandas as pd
from sklearn import datasets
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.feature_extraction.text import CountVectorizer
from xgboost import XGBClassifier,XGBRegressor
from sklearn_pandas import DataFrameMapper
from sklearn.model_selection import train_test_split

auto = pd.read_csv('auto-mpg.csv')
X = auto.drop(['mpg', 'car name'], axis=1)
y = auto['mpg']

x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=101)

model = XGBRegressor(random_state=1, seed=1)

model.fit(x_train, y_train)

Predictions accuracy: 

![image](https://user-images.githubusercontent.com/47885698/77522366-1117e480-6e7c-11ea-811d-e28995138db5.png)


Model when wrapped in your wrapped and reduced predictive accuracy: 

auto = pd.read_csv('auto-mpg.csv')
X = auto.drop(['mpg', 'car name'], axis=1)
y = auto['mpg']

feature_names = [name for name in auto.columns if name not in ('mpg')]

target_name='mpg'


x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=101)
pipeline_obj = Pipeline([
    ('mapper', DataFrameMapper([
        (['displacement'],[StandardScaler()]) 
    ])),
    ('Xgbr',XGBRegressor(random_state=1, seed=1))
])
pipeline_obj.fit(x_train,y_train)

![image](https://user-images.githubusercontent.com/47885698/77522471-360c5780-6e7c-11ea-9b5c-4d1d6fc954ac.png)



We shouldn't expect the model when put into a wrapper to be less accurate surely?

**Comment by 1samwatkins3005** [comment]
I've fixed this locally. My mistake, not package mistake.
